pageserver/src/layered_repository/image_layer.rs

```diff
@@ -385,16 +385,37 @@ impl ImageLayer {
     }
 
     ///
-    /// Load the contents of the file into memory
+    /// Open the underlying file and read the metadata into memory, if it's
+    /// not loaded already.
     ///
-    fn load(&self) -> Result<MutexGuard<ImageLayerInner>> {
-        // quick exit if already loaded
-        let mut inner = self.inner.lock().unwrap();
+    fn load(&self) -> Result<RwLockReadGuard<ImageLayerInner>> {
+        loop {
+            // Quick exit if already loaded
+            let inner = self.inner.read().unwrap();
+            if inner.book.is_some() {
+                return Ok(inner);
+            }
+
+            // Need to open the file and load the metadata. Upgrade our lock to
+            // a write lock. (Or rather, release and re-lock in write mode.)
+            drop(inner);
+            let inner = self.inner.write().unwrap();
+            if inner.book.is_none() {
+                self.load_inner(inner)?;
+            } else {
+                // Another thread loaded it while we were not holding the lock.
+            }
 
-        if inner.book.is_some() {
-            return Ok(inner);
+            // We now have the file open and loaded. There's no function to do
+            // that in the std library RwLock, so we have to release and re-lock
+            // in read mode. (To be precise, the lock guard was moved in the
+            // above call to `load_inner`, so it's already been released). And
+            // while we do that, another thread could unload again, so we have
+            // to re-check and retry if that happens.
```

Comment:
If we keep using std::sync for locks, then an explicit drop(inner) would probably help with understanding the code at a glance, so that we don't need to rely on the fact that the variable's lifetime ends at the end of the loop and thus the lock is dropped as well.

Reply:
I switched to use parking_lot synchronization primitives in buffered_repository branch: it provides some small increase of performance, but what is more important: significantly reduced latency. Even normal (not fair) mutex/rwlock prevents starvation.

Also lack of wrap() and poisoned state is nice feature.
There are also some advanced features like recursive locks and downgrade/upgrade but I have not used them.

Reply:
But std::sync primitives have one important advantage: them are much easier for debug. Them are using pthread library function call and it is possible to dump in debugger for example rwock object - it shows number of readers and thread ID of owner (if any).
I failed to extract this information from parting_lot objects...

Reply:
Added an explicit `drop(inner)` here per @MMeent's suggestion.